consensus/spos/worker.go

```diff
@@ -711,6 +720,112 @@ func (wrk *Worker) ResetConsensusMessages() {
 	wrk.consensusMessageValidator.resetConsensusMessages()
 }
 
+func (wrk *Worker) checkValidityAndProcessEquivalentMessages(cnsMsg *consensus.Message, p2pMessage p2p.MessageP2P) error {
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+
+	log.Trace("received message from consensus topic",
+		"msg type", wrk.consensusService.GetStringValue(msgType),
+		"from", cnsMsg.PubKey,
+		"header hash", cnsMsg.BlockHeaderHash,
+		"round", cnsMsg.RoundIndex,
+		"size", len(p2pMessage.Data()),
+	)
+
+	if !wrk.enableEpochsHandler.IsFlagEnabled(common.EquivalentMessagesFlag) {
+		return wrk.consensusMessageValidator.checkConsensusMessageValidity(cnsMsg, p2pMessage.Peer())
+	}
+
+	// if the message is not with final info, no need to check its equivalent messages
+	if !wrk.consensusService.IsMessageWithFinalInfo(msgType) {
+		return wrk.consensusMessageValidator.checkConsensusMessageValidity(cnsMsg, p2pMessage.Peer())
+	}
+
+	wrk.mutEquivalentMessages.Lock()
+	defer wrk.mutEquivalentMessages.Unlock()
+
+	err := wrk.processEquivalentMessageUnprotected(cnsMsg)
+	if err != nil {
+		return err
+	}
+
+	err = wrk.consensusMessageValidator.checkConsensusMessageValidity(cnsMsg, p2pMessage.Peer())
+	if err != nil {
+		wrk.processInvalidEquivalentMessageUnprotected(cnsMsg.BlockHeaderHash)
+		return err
+	}
+
+	return nil
+}
+
+func (wrk *Worker) processEquivalentMessageUnprotected(cnsMsg *consensus.Message) error {
+	err := wrk.verifyEquivalentMessageSignature(cnsMsg)
+	if err != nil {
+		return err
+	}
+
+	hdrHash := string(cnsMsg.BlockHeaderHash)
+
+	// if a valid equivalent message was seen before, return error to stop further broadcasts
+	numMessages := wrk.equivalentMessages[hdrHash]
+	wrk.equivalentMessages[hdrHash] = numMessages + 1
+	if numMessages > 0 {
+		return ErrEquivalentMessageAlreadyReceived
+	}
+
+	return nil
+}
+
+func (wrk *Worker) verifyEquivalentMessageSignature(cnsMsg *consensus.Message) error {
+	if check.IfNil(wrk.consensusState.Header) {
+		return ErrNilHeader
+	}
+
+	header := wrk.consensusState.Header.ShallowClone()
+
+	err := header.SetSignature(cnsMsg.Signature)
```

Comment:
did we agree to only temporarily set this on the header and remove it afterwards?
otherwise the header sig verifier should accept separately the header and the proof (bitmap + aggregated signature) and then verify

Reply:
removed and added todo for next PR, where the header structure is updated